src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectShimFactory.cs

```diff
@@ -0,0 +1,88 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices.Implementation.TaskList;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    [Export(typeof(IProjectShimFactory))]
+    internal partial class ProjectShimFactory : IProjectShimFactory
+    {
+        private readonly VisualStudioProjectTracker _projectTracker;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly VisualStudioWorkspaceImpl _visualStudioWorkspace;
+        private readonly HostDiagnosticUpdateSource _hostDiagnosticUpdateSource;
+        private readonly Dictionary<string, IVsReportExternalErrors> _externalErrorReporterMap;
+        private readonly ImmutableDictionary<string, string> _projectLangaugeToErrorCodePrefixMap =
+            ImmutableDictionary.CreateRange(StringComparer.OrdinalIgnoreCase, new[]
+            {
+                new KeyValuePair<string, string> (LanguageNames.CSharp, "CS"),
+                new KeyValuePair<string, string> (LanguageNames.VisualBasic, "BC"),
+            });
+
+        [ImportingConstructor]
+        public ProjectShimFactory(
+            VisualStudioProjectTracker projectTracker,
+            SVsServiceProvider serviceProvider,
+            VisualStudioWorkspaceImpl visualStudioWorkspace,
+            HostDiagnosticUpdateSource hostDiagnosticUpdateSource)
+        {
+            _projectTracker = projectTracker;
+            _serviceProvider = serviceProvider;
+            _visualStudioWorkspace = visualStudioWorkspace;
+            _hostDiagnosticUpdateSource = hostDiagnosticUpdateSource;
+            _externalErrorReporterMap = new Dictionary<string, IVsReportExternalErrors>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        // internal for testing purposes only.
+        internal static IProjectShim CreateProjectShim(VisualStudioProjectTracker projectTracker, IServiceProvider serviceProvider, IVsHierarchy hierarchy, string projectFilePath, string language, Guid projectGuid, CommandLineArguments commandLineArguments)
+        {
+            return new ProjectShim(commandLineArguments, projectTracker, reportExternalErrorCreatorOpt: null, hierarchy: hierarchy, language: language,
+                serviceProvider: serviceProvider, visualStudioWorkspaceOpt: null, hostDiagnosticUpdateSourceOpt: null, projectFilePath: projectFilePath, projectGuid: projectGuid);
+        }
+
+        IProjectShim IProjectShimFactory.CreateProjectShim(string languageName, string projectFilePath, Guid projectGuid, object hostObject, CommandLineArguments commandLineArguments)
+        {
+            var vsHierarchy = hostObject as IVsHierarchy;
```

Comment:
Why isn't there a little more validation to help the caller here, when:

1) The language is not recognized
2) You pass something other than IVsHierarchy as the "hostObject" for now